caas/kubernetes/provider/cloud.go

```diff
@@ -92,61 +91,97 @@ func newCloudCredentialFromKubeConfig(reader io.Reader, cloudParams KubeCloudPar
+
+			// Record the operator storage to use.
+			if clusterMetadata.OperatorStorageClass != nil {
+				operatorSC = clusterMetadata.OperatorStorageClass.Name
+				storageMsg += "."
+			} else {
+				if storageMsg == "" {
+					storageMsg += "\nwith "
+				} else {
+					storageMsg += "\n"
+				}
+				storageMsg += fmt.Sprintf("operator storage provisioned by the workload storage class.")
+			}
+
+			if clusterMetadata.NominatedStorageClass != nil {
+				workloadSC = clusterMetadata.NominatedStorageClass.Name
+			}
+
+			if k8sCloud.Config == nil {
+				k8sCloud.Config = make(map[string]interface{})
+			}
+			if _, ok := k8sCloud.Config[WorkloadStorageKey]; !ok {
+				k8sCloud.Config[WorkloadStorageKey] = workloadSC
+			}
+			if _, ok := k8sCloud.Config[OperatorStorageKey]; !ok {
+				k8sCloud.Config[OperatorStorageKey] = operatorSC
+			}
+		}
+	}()
 
 	if err != nil || clusterMetadata == nil {
 		// err will be nil if user hit Ctrl+C.
 		msg := "cannot get cluster metadata"
 		if err != nil {
 			msg = err.Error()
 		}
-		return fail(ClusterQueryError{Message: msg})
+		return "", ClusterQueryError{Message: msg}
 	}
 
-	if storageParams.HostCloudRegion == "" && clusterMetadata.Regions != nil && clusterMetadata.Regions.Size() > 0 {
-		storageParams.HostCloudRegion = cloud.BuildHostCloudRegion(
-			clusterMetadata.Cloud,
-			clusterMetadata.Regions.SortedValues()[0],
-		)
+	if storageParams.HostCloudRegion == "" && clusterMetadata.Cloud != "" {
+		var region string
+		if clusterMetadata.Regions != nil && clusterMetadata.Regions.Size() > 0 {
+			region = clusterMetadata.Regions.SortedValues()[0]
+		}
+		storageParams.HostCloudRegion = cloud.BuildHostCloudRegion(clusterMetadata.Cloud, region)
 	}
-	if storageParams.HostCloudRegion == "" {
-		return fail(ClusterQueryError{})
+	k8sCloud.HostCloudRegion = storageParams.HostCloudRegion
+
+	cloudType, region, err := cloud.SplitHostCloudRegion(k8sCloud.HostCloudRegion)
+	if err == nil {
+		// try to get region if possible.
+		k8sCloud.Regions = []cloud.Region{{
+			Name: region,
+		}}
 	}
-	_, region, err := ParseCloudRegion(storageParams.HostCloudRegion)
-	if err != nil {
-		return fail(errors.Annotatef(err, "validating cloud region %q", storageParams.HostCloudRegion))
+	if clusterMetadata.NominatedStorageClass != nil && clusterMetadata.OperatorStorageClass != nil {
+		// all good.
+		return "", nil
 	}
-	k8sCloud.HostCloudRegion = storageParams.HostCloudRegion
-	k8sCloud.Regions = []cloud.Region{{
-		Name: region,
-	}}
 
-	// If the user has not specified storage, check that the cluster has Juju's opinionated defaults.
-	cloudType := cloud.SplitHostCloudRegion(storageParams.HostCloudRegion)[0]
-	err = storageParams.MetadataChecker.CheckDefaultWorkloadStorage(cloudType, clusterMetadata.NominatedStorageClass)
+	if err != nil {
+		// Region is optional, but cloudType is required for next step.
+		return "", ClusterQueryError{}
+	}
 
+	// check Juju's opinionated defaults if cloudType is usable.
```

Comment:
Can we retain the original comment
// If the user has not specified storage, check ...

Reply:
fixed